### ISSUE TITLE
avoid_function_literals_in_foreach_calls fails to handle legacy types

### ISSUE THREAD

**Issue by fishythefish** [dart-lang/linter#4328]
Culprit:

https://github.com/dart-lang/linter/blob/997671c8c67e3f8b7ce6a0e379fc308b5636df07/lib/src/rules/avoid_function_literals_in_foreach_calls.dart#L100

This check doesn't allow for a `star` nullability suffix:

https://github.com/dart-lang/linter/blob/997671c8c67e3f8b7ce6a0e379fc308b5636df07/lib/src/rules/avoid_function_literals_in_foreach_calls.dart#L59-L62

But that means that code will suddenly start violating this lint when it's migrated.

**Comment by fishythefish** [comment]
Here's a simple repro.

My `analysis_options.yaml`:
```yaml
linter:
  rules:
    - avoid_function_literals_in_foreach_calls
```

Now consider the following unmigrated file:

```dart
// ignore: illegal_language_version_override
// @dart=2.11

void main() {
  List<int> x = [1, 2, 3];
  x.forEach((n) {
    print(n);
  });
}
```

If I run `dart analyze`, I get no errors. I now migrate the file:

```dart
void main() {
  List<int> x = [1, 2, 3];
  x.forEach((n) {
    print(n);
  });
}
```

All of a sudden, I get a lint violation.

**Comment by fishythefish** [comment]
Gut feeling is that the fix is to `|| type.nullabilitySuffix == NullabilitySuffix.star` (which I think means we can actually ignore the suffix completely?) since we already distinguish between `.` and `?.` elsewhere. But I haven't thought about edge cases at all.

**Comment by oprypin** [comment]
Hm I think it's true,

* https://github.com/dart-lang/linter/pull/4039

this PR aimed to fix one thing but caused this big unrelated change.

Since then I have made a further change to this lint, which I'm pretty sure will keep this added test passing but without the older code change.

* #4305

But actually now that I think about it, this was probably the PR that caused the sequence of confusing events :)

* #2752

Whether the iterable can be null is not the point because if you're writing `foo.forEach`, it's anyway fully equivalent to replace it with a for-loop, they'll have run-time errors in the same cases.

So in summary I think `_isNonNullableIterable` can just be changed back to the check whether it implements `Iterable`.

**Comment by oprypin** [comment]
Linked a PR accordingly
